test/cpp/naming/naming_end2end_test.cc

```diff
@@ -0,0 +1,297 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/host_port.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+#include <grpc/support/time.h>
+#include <string.h>
+
+#include <gflags/gflags.h>
+#include <vector>
+
+#include "test/cpp/util/test_config.h"
+
+extern "C" {
+#include "src/core/ext/filters/client_channel/client_channel.h"
+#include "src/core/ext/filters/client_channel/resolver.h"
+#include "src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h"
+#include "src/core/ext/filters/client_channel/resolver_registry.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/iomgr/combiner.h"
+#include "src/core/lib/iomgr/executor.h"
+#include "src/core/lib/iomgr/iomgr.h"
+#include "src/core/lib/iomgr/resolve_address.h"
+#include "src/core/lib/iomgr/sockaddr_utils.h"
+#include "src/core/lib/support/env.h"
+#include "src/core/lib/support/string.h"
+#include "test/core/util/test_config.h"
+}
+
+using std::vector;
+
+DEFINE_string(target_name, "", "");
+DEFINE_bool(expect_target_is_balancer, true, "");
+DEFINE_string(expected_addrs, "", "");
+DEFINE_string(expected_chosen_service_config, "", "");
+DEFINE_string(local_dns_server_address, "", "");
+
+namespace grpc {
+namespace testing {
+
+static vector<std::string> parse_expected_addrs(const char *expected_addrs) {
+  std::vector<std::string> out;
+
+  char *p = (char *)expected_addrs;
+
+  if (strlen(p) != 0) {
+    for (;;) {
+      char *comma = strchr(p, ',');
+      if (comma) {
+        out.emplace_back(p, comma);
+        p = comma + 1;
+      } else {
+        out.emplace_back(p);
+        break;
+      }
+    }
+  }
+  if (out.size() == 0) {
+    gpr_log(GPR_ERROR,
+            "expected_addrs arg should be a comma-separated list of ip-ports");
+  }
+  return out;
+}
+
+static gpr_timespec test_deadline(void) {
+  return grpc_timeout_seconds_to_deadline(100);
+}
+
+typedef struct args_struct {
+  gpr_event ev;
+  gpr_atm done_atm;
+  gpr_mu *mu;
+  grpc_pollset *pollset;
+  grpc_pollset_set *pollset_set;
+  grpc_combiner *lock;
+  grpc_channel_args *channel_args;
+  bool expect_is_balancer;
+  const char *target_name;
+  vector<std::string> expected_addrs;
+  const char *expected_service_config_string;
+} args_struct;
+
+int matches_any(vector<std::string> expected_addrs, const char *addr) {
+  for (auto it = expected_addrs.begin(); it != expected_addrs.end(); it++) {
+    if (it->compare(addr) == 0) {
```

Comment:
If you pass in `addr` as a `std::string` instead of a `const char*`, you can just use the `==` operator here.